### PR TITLE
adds support for deleting init node

### PR DESCRIPTION
**Pull request by Josh-Diamond**
extensions to support: https://github.com/rancher/rancher/pull/44155

### CODE REVIEW

File: extensions/nodes/nodes.go
Comment (igomez06) [inline extensions/nodes/nodes.go:88]: So it's delete a cluster.x-k8s.io.machine not the node?

File: extensions/nodes/nodes.go
Comment (igomez06) [inline extensions/nodes/nodes.go:97]: Do we really need the verify here? Shouldn't this just be in the test?

**Review by igomez06 — COMMENTED** [review]
A few questions.

File: extensions/nodes/nodes.go
Comment (Josh-Diamond) [reply extensions/nodes/nodes.go:97]: I felt this logic was re-usable and could be useful in future tests, so in an effort to keep things DRY moving forward, I broke it out into an extension.  Upon reviewing this, I'm noticing the logs are specific to my `init node` task, so I will need to update them to be more generic, unless its requested to remove this from the extensions altogether.

File: extensions/nodes/nodes.go
Comment (Josh-Diamond) [reply extensions/nodes/nodes.go:88]: Yes.  The function accepts a v3.Node object, but its essentially targeting and deleting a machine.

File: extensions/nodes/nodes.go
Comment (markusewalker) [inline extensions/nodes/nodes.go:22]: nit - can the comment just say `GetInitNode` and not `GetInitNode()`.

File: extensions/nodes/nodes.go
Comment (markusewalker) [inline extensions/nodes/nodes.go:79]: Same nit here.

File: extensions/nodes/nodes.go
Comment (markusewalker) [reply extensions/nodes/nodes.go:88]: This function as well is more generic as opposed to searching for the init node specifically.

File: extensions/nodes/nodes.go
Comment (markusewalker) [inline extensions/nodes/nodes.go:96]: Same nit here.

**Review by markusewalker — COMMENTED** [review]
Overall, looks good. Left few comments.

File: extensions/nodes/nodes.go
Comment (Josh-Diamond) [reply extensions/nodes/nodes.go]: Done!

File: extensions/nodes/nodes.go
Comment (Josh-Diamond) [reply extensions/nodes/nodes.go]: Done!

File: extensions/nodes/nodes.go
Comment (Josh-Diamond) [reply extensions/nodes/nodes.go:96]: Done!

**Review by slickwarren — APPROVED** [review]
lgtm

**Review by markusewalker — COMMENTED** [review]
Overall, LGTM. Can you make sure that you run this PR against the linter the team discussed and adjust accordingly to its feedback?

Also, once this PR is approved, please open the backport PRs.

File: extensions/nodes/nodes.go
Comment (caliskanugur) [inline extensions/nodes/nodes.go:50]: A question here, does it need to create nested for loops? It can check map keys, something like this;

```suggestion
	for _, machinePlan := range machinePlans {
	        v, ok := machinePlan.ObjectMeta.Labels[initNodeLabelKey]
	        if ok && v ==  initNodeLabelValue {
	             		logrus.Infof("Found init node machine plan: %s", machinePlan.Name)
				initNodeMachineName = machinePlan.Labels[machineNameSteveLabel]
		}
	}
```

File: extensions/nodes/nodes.go
Comment (caliskanugur) [inline extensions/nodes/nodes.go:67]: The same applies here too, could you check if the map contains a key rather than creating a loop for the map?

File: extensions/nodes/nodes.go
Comment (igomez06) [inline extensions/nodes/nodes.go:108]: Why is this a fatal. We should never have fatals ever. It completely exits the test, all tests cancel. We don't want that.

File: extensions/nodes/nodes.go
Comment (igomez06) [inline extensions/nodes/nodes.go:100]: My only other nit, is new code should use v1 calls and stop using v3/norman. Unless you can convince me otherwise.

File: extensions/nodes/nodes.go
Comment (igomez06) [reply extensions/nodes/nodes.go:88]: Why does it accept a v3.node instead of a v1 node?

**Review by caliskanugur — COMMENTED** [review]
I've got a few questions

File: extensions/nodes/nodes.go
Comment (caliskanugur) [inline extensions/nodes/nodes.go:15]: I think this is not necessary it can be a parse `b, err := strconv.ParseBool("true")` something like this once it gets the label? This would happen in the code snippet I commented below

File: extensions/nodes/nodes.go
Comment (Josh-Diamond) [reply extensions/nodes/nodes.go]: I've updated this to use v1

File: extensions/nodes/nodes.go
Comment (Josh-Diamond) [reply extensions/nodes/nodes.go]: I removed the fatal

File: extensions/nodes/nodes.go
Comment (Josh-Diamond) [reply extensions/nodes/nodes.go]: great suggestion! thank you Ugur! i've implemented this

**Review by markusewalker — COMMENTED** [review]
Linter caught two issues. Will approve once they have been addressed.

**Review by markusewalker — APPROVED** [review]
LGTM

**Review by susesgartner — APPROVED** [review]
LGTM

**Review by slickwarren — APPROVED** [review]

File: extensions/nodes/nodes.go
Comment (caliskanugur) [inline extensions/nodes/nodes.go:45]: Josh, could you please check this: https://github.com/rancher/shepherd/pull/64#discussion_r1525497184. You can find more info on offline PR guidelines docs (15th)

Comment (Josh-Diamond) [reply extensions/nodes/nodes.go:45]: Hey Ugur,

I reviewed the comment, and when implementing it, I was getting a type error where bool cannot be type string, so I hardcoded this to "true" as it should never change. I felt this resolved the initial concerns with using the const and is a simple and clean resolution.

Do you have any concerns with this approach? 🙂

**Pull request closed without merging**
